File: No_1339/Maximum Product of Splitted Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    long long int total = 0;
    long long int ret = 0;
public:
    int maxProduct(TreeNode* root) {
        total = TreeSum(root);
        solve(root);
        return ret % (int)(1e9+7);
    }
    
    long long int TreeSum(TreeNode* root) {
        if(root == nullptr) return 0;
        return root->val + TreeSum(root->left) + TreeSum(root->right);
    }
    
    long long int solve(TreeNode* root) {
        if(root == nullptr) return 0;
        long long int left = solve(root->left);
        long long int right = solve(root->right);
        if(left != 0) {
            ret = max(ret, left*(total-left));
        }
        if(right != 0) {
            ret = max(ret, right*(total-right));
        }
        
        return left + right + root->val;
    }
};

File: No_1339/Maximum Product of Splitted Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        MOD = 10**9 + 7
        max_product = 0

        def compute_total_sum(node: Optional[TreeNode]) -> int:
            if not node:
                return 0
            return (
                node.val
                + compute_total_sum(node.left)
                + compute_total_sum(node.right)
            )

        def postorder_sum(node: Optional[TreeNode]) -> int:
            nonlocal max_product
            if not node:
                return 0

            subtree_sum = (
                node.val
                + postorder_sum(node.left)
                + postorder_sum(node.right)
            )

            max_product = max(
                max_product,
                subtree_sum * (total_sum - subtree_sum)
            )

            return subtree_sum

        total_sum = compute_total_sum(root)
        postorder_sum(root)

        return max_product % MOD

        